app/ide-desktop/lib/dashboard/src/authentication/src/hooks.tsx

```diff
@@ -0,0 +1,56 @@
+/** @file Module containing common custom React hooks used throughout out Dashboard. */
+import * as React from "react";
+import * as loggerProvider from "./providers/logger";
+
+
+
+// ======================
+// === useAsyncEffect ===
+// ======================
+
+/** A React hook for re-rendering a component once an asynchronous call is over.
+ *
+ * This hook will take care of setting an initial value for the component state (so that it can
+ * render immediately), updating the state once the asynchronous call is over (to re-render the
+ * component), and cancelling any in-progress asynchronous calls when the component is unmounted (to
+ * avoid race conditions where "update 1" starts, "update 2" starts and finishes, then "update 1"
+ * finishes and sets the state).
+ *
+ * For further details, see: https://devtrium.com/posts/async-functions-useeffect
+ * Also see: https://stackoverflow.com/questions/61751728/asynchronous-calls-with-react-usememo
+ *
+ * @param initialValue - The initial value of the state controlled by this hook.
+ * @param fetch - The asynchronous function used to load the state controlled by this hook.
+ * @param deps - The list of dependencies that, when updated, trigger the asynchronous fetch.
+ * @returns value - The current value of the state controlled by this hook. */
+export function useAsyncEffect<T>(
+    initialValue: T,
+    fetch: () => Promise<T>,
+    deps?: React.DependencyList
+): [T] {
```

Comment:
is there a reason the return type isn't just `: T`? returning `[T]` *would* trigger an update every time this is called, but is that desirable? (and if it is, maybe documentation should be added explaining why it must be like this)

Reply:
Good catch. It's `[T]` because this function was returning more than one value, as an array. It was like `[value, setValue]` or something. This can just be `T` now